stress: failed test in cockroach/gossip/gossip.test: TestClientNodeID

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/e9f364bea3014b627b66e8dddcc79a4f3cbcfaad

Stress build found a failed test:

```
=== RUN   TestClientNodeID
W160311 03:50:09.218659 gossip/gossip.go:912  not connected to cluster; use --join to specify a connected node
E160311 03:50:09.224185 rpc/context.go:146  rpc error: code = 13 desc = "transport is closing"
--- FAIL: TestClientNodeID (5.03s)
	leaktest.go:80: Leaked goroutine: goroutine 178 [select]:
		google.golang.org/grpc/transport.(*serverHandlerTransport).runStream(0xc8200106c0)
			/go/src/google.golang.org/grpc/transport/handler_server.go:345 +0x115
		google.golang.org/grpc/transport.(*serverHandlerTransport).HandleStreams(0xc8200106c0, 0xc820177ce0)
			/go/src/google.golang.org/grpc/transport/handler_server.go:335 +0xca3
		google.golang.org/grpc.(*Server).serveStreams(0xc820096480, 0x7f00771afbe0, 0xc8200106c0)
			/go/src/google.golang.org/grpc/server.go:325 +0x159
		google.golang.org/grpc.(*Server).ServeHTTP(0xc820096480, 0x7f00771afb58, 0xc820120040, 0xc8201661c0)
			/go/src/google.golang.org/grpc/server.go:368 +0x1a8
		net/http.(Handler).ServeHTTP-fm(0x7f00771afb58, 0xc820120040, 0xc8201661c0)
			/usr/local/go/src/net/http/h2_bundle.go:3847 +0x50
		golang.org/x/net/http2.(*serverConn).runHandler(0xc82017e1e0, 0xc820120040, 0xc8201661c0, 0xc82013c240)
			/go/src/golang.org/x/net/http2/server.go:1649 +0x9f
		created by golang.org/x/net/http2.(*serverConn).processHeaders
			/go/src/golang.org/x/net/http2/server.go:1428 +0x8d8
```

Run Details:

```
63 runs so far, 0 failures, over 5s
128 runs so far, 0 failures, over 10s
182 runs so far, 0 failures, over 15s
197 runs completed, 1 failures, over 16s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client_test.go
+++ b/gossip/client_test.go
@@ -42,11 +42,7 @@ func startGossip(nodeID roachpb.NodeID, stopper *stop.Stopper, t *testing.T) *Go
 
 	addr := util.CreateTestAddr("tcp")
 	server := rpc.NewServer(rpcContext)
-	tlsConfig, err := rpcContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	ln, err := util.ListenAndServe(stopper, server, addr, tlsConfig)
+	ln, err := util.ListenAndServeGRPC(stopper, server, addr)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -106,11 +102,7 @@ func startFakeServerGossips(t *testing.T) (local *Gossip, remote *fakeGossipServ
 
 	laddr := util.CreateTestAddr("tcp")
 	lserver := rpc.NewServer(lRPCContext)
-	lTLSConfig, err := lRPCContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	lln, err := util.ListenAndServe(stopper, lserver, laddr, lTLSConfig)
+	lln, err := util.ListenAndServeGRPC(stopper, lserver, laddr)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -121,11 +113,7 @@ func startFakeServerGossips(t *testing.T) (local *Gossip, remote *fakeGossipServ
 
 	raddr := util.CreateTestAddr("tcp")
 	rserver := rpc.NewServer(rRPCContext)
-	rTLSConfig, err := rRPCContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	rln, err := util.ListenAndServe(stopper, rserver, raddr, rTLSConfig)
+	rln, err := util.ListenAndServeGRPC(stopper, rserver, raddr)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -133,7 +121,6 @@ func startFakeServerGossips(t *testing.T) (local *Gossip, remote *fakeGossipServ
 	remote = newFakeGossipServer(rserver, stopper)
 	addr := rln.Addr()
 	remote.nodeAddr = util.MakeUnresolvedAddr(addr.Network(), addr.String())
-	time.Sleep(time.Millisecond)
 	return
 }
 
@@ -324,11 +311,7 @@ func TestClientRegisterWithInitNodeID(t *testing.T) {
 
 		addr := util.CreateTestAddr("tcp")
 		server := rpc.NewServer(RPCContext)
-		TLSConfig, err := RPCContext.GetServerTLSConfig()
-		if err != nil {
-			t.Fatal(err)
-		}
-		ln, err := util.ListenAndServe(stopper, server, addr, TLSConfig)
+		ln, err := util.ListenAndServeGRPC(stopper, server, addr)
 		if err != nil {
 			t.Fatal(err)
 		}

--- a/gossip/simulation/network.go
+++ b/gossip/simulation/network.go
@@ -89,7 +89,7 @@ func NewNetwork(nodeCount int) *Network {
 func (n *Network) CreateNode() (*Node, error) {
 	server := rpc.NewServer(n.rpcContext)
 	testAddr := util.CreateTestAddr("tcp")
-	ln, err := util.ListenAndServe(n.Stopper, server, testAddr, n.tlsConfig)
+	ln, err := util.ListenAndServeGRPC(n.Stopper, server, testAddr)
 	if err != nil {
 		return nil, err
 	}

--- a/kv/send_test.go
+++ b/kv/send_test.go
@@ -49,18 +49,8 @@ func newNodeTestContext(clock *hlc.Clock, stopper *stop.Stopper) *rpc.Context {
 func newTestServer(t *testing.T, ctx *rpc.Context) (*grpc.Server, net.Listener) {
 	s := rpc.NewServer(ctx)
 
-	tlsConfig, err := ctx.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// We may be called in a loop, meaning tlsConfig is may be in used by a
-	// running server during a call to `util.ListenAndServe`, which may
-	// mutate it (due to http2.ConfigureServer). Make a copy to avoid trouble.
-	tlsConfigCopy := *tlsConfig
-
 	addr := util.CreateTestAddr("tcp")
-	ln, err := util.ListenAndServe(ctx.Stopper, s, addr, &tlsConfigCopy)
+	ln, err := util.ListenAndServeGRPC(ctx.Stopper, s, addr)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/rpc/context_test.go
+++ b/rpc/context_test.go
@@ -21,6 +21,7 @@ import (
 	"testing"
 
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
 
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
@@ -29,15 +30,14 @@ import (
 )
 
 func newTestServer(t *testing.T, ctx *Context, manual bool) (*grpc.Server, net.Listener) {
-	s := grpc.NewServer()
-
 	tlsConfig, err := ctx.GetServerTLSConfig()
 	if err != nil {
 		t.Fatal(err)
 	}
+	s := grpc.NewServer(grpc.Creds(credentials.NewTLS(tlsConfig)))
 
 	addr := util.CreateTestAddr("tcp")
-	ln, err := util.ListenAndServe(ctx.Stopper, s, addr, tlsConfig)
+	ln, err := util.ListenAndServeGRPC(ctx.Stopper, s, addr)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/server/node_test.go
+++ b/server/node_test.go
@@ -61,11 +61,7 @@ func createTestNode(addr net.Addr, engines []engine.Engine, gossipBS net.Addr, t
 	ctx.ScanInterval = 10 * time.Hour
 	ctx.ConsistencyCheckInterval = 10 * time.Hour
 	grpcServer := rpc.NewServer(nodeRPCContext)
-	tlsConfig, err := nodeRPCContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	ln, err := util.ListenAndServe(stopper, grpcServer, addr, tlsConfig)
+	ln, err := util.ListenAndServeGRPC(stopper, grpcServer, addr)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -544,15 +544,11 @@ func (m *multiTestContext) addStore() {
 		}
 	}
 
-	tlsConfig, err := m.rpcContext.GetServerTLSConfig()
-	if err != nil {
-		m.t.Fatal(err)
-	}
-
 	if m.nodeIDtoAddr == nil {
 		m.nodeIDtoAddr = make(map[roachpb.NodeID]net.Addr)
 	}
-	ln, err := util.ListenAndServe(m.transportStopper, m.grpcServers[idx], util.CreateTestAddr("tcp"), tlsConfig)
+	ln, err := util.ListenAndServeGRPC(m.transportStopper,
+		m.grpcServers[idx], util.CreateTestAddr("tcp"))
 	if err != nil {
 		m.t.Fatal(err)
 	}

--- a/storage/raft_transport_test.go
+++ b/storage/raft_transport_test.go
@@ -91,22 +91,12 @@ func TestSendAndReceive(t *testing.T) {
 		5: 1,
 	}
 
-	tlsConfig, err := nodeRPCContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	// tls.Config is not safe for concurrent use because it is lazily initialized
-	// - grpc/credentials.NewTLS (called from rpc.NewServer) makes a shallow copy
-	// of the passed-in *tls.Config which can race with this lazy initialization.
-	// To work around this, we make a single copy here which will be used by all
-	// HTTP servers and not shared with calls to grpc/credentials.NewTLS.
-	tlsConfigCopy := *tlsConfig
-
 	for serverIndex := 0; serverIndex < numServers; serverIndex++ {
 		nodeID := nextNodeID
 		nextNodeID++
 		grpcServer := rpc.NewServer(nodeRPCContext)
-		ln, err := util.ListenAndServe(stopper, grpcServer, util.CreateTestAddr("tcp"), &tlsConfigCopy)
+		ln, err := util.ListenAndServeGRPC(stopper, grpcServer,
+			util.CreateTestAddr("tcp"))
 		if err != nil {
 			t.Fatal(err)
 		}
@@ -122,7 +112,8 @@ func TestSendAndReceive(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		transport := storage.NewRaftTransport(storage.GossipAddressResolver(g), grpcServer, nodeRPCContext)
+		transport := storage.NewRaftTransport(storage.GossipAddressResolver(g),
+			grpcServer, nodeRPCContext)
 		transports[nodeID] = transport
 
 		for store := 0; store < storesPerServer; store++ {
@@ -241,11 +232,7 @@ func TestInOrderDelivery(t *testing.T) {
 	g := gossip.New(nodeRPCContext, nil, stopper)
 
 	grpcServer := rpc.NewServer(nodeRPCContext)
-	tlsConfig, err := nodeRPCContext.GetServerTLSConfig()
-	if err != nil {
-		t.Fatal(err)
-	}
-	ln, err := util.ListenAndServe(stopper, grpcServer, util.CreateTestAddr("tcp"), tlsConfig)
+	ln, err := util.ListenAndServeGRPC(stopper, grpcServer, util.CreateTestAddr("tcp"))
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/util/leaktest/leaktest.go
+++ b/util/leaktest/leaktest.go
@@ -32,7 +32,7 @@ func interestingGoroutines() (gs []string) {
 
 		if stack == "" ||
 			strings.Contains(stack, "github.com/cockroachdb/cockroach/util/log.init") ||
-
+			// Below are the stacks ignored by the upstream leaktest code.
 			strings.Contains(stack, "testing.Main(") ||
 			strings.Contains(stack, "runtime.goexit") ||
 			strings.Contains(stack, "created by runtime.gc") ||

--- a/util/net.go
+++ b/util/net.go
@@ -25,6 +25,8 @@ import (
 	"sync"
 	"time"
 
+	"google.golang.org/grpc"
+
 	"golang.org/x/net/http2"
 
 	"github.com/cockroachdb/cmux"
@@ -32,60 +34,23 @@ import (
 	"github.com/cockroachdb/cockroach/util/stop"
 )
 
-// ListenAndServe creates a listener and serves handler on it, closing the
-// listener when signalled by the stopper. The handling server implements HTTP1
-// and HTTP2, with or without TLS. Note that the "real" server also implements
-// the postgres wire protocol, and so does not use this function, but the
-// pattern used is similar; that implementation is in server/server.go.
-func ListenAndServe(stopper *stop.Stopper, handler http.Handler, addr net.Addr, tlsConfig *tls.Config) (net.Listener, error) {
+// ListenAndServeGRPC creates a listener and serves the specified grpc Server
+// on it, closing the listener when signalled by the stopper.
+func ListenAndServeGRPC(stopper *stop.Stopper, server *grpc.Server,
+	addr net.Addr) (net.Listener, error) {
 	ln, err := net.Listen(addr.Network(), addr.String())
 	if err != nil {
 		return ln, err
 	}
+
 	stopper.RunWorker(func() {
 		<-stopper.ShouldDrain()
-		// Some unit tests manually close `ln`, so it may already be closed
-		// when we get here.
-		FatalIfUnexpected(ln.Close())
+		server.Stop()
 	})
 
-	if tlsConfig != nil {
-		// We're in TLS mode. ALPN will be used to automatically handle HTTP1 and
-		// HTTP2 requests.
-		ServeHandler(stopper, handler, tls.NewListener(ln, tlsConfig), tlsConfig)
-	} else {
-		// We're not in TLS mode. We're going to implement h2c (HTTP2 Clear Text)
-		// ourselves.
-
-		m := cmux.New(ln)
-		// HTTP2 connections are easy to identify because they have a common
-		// preface.
-		h2L := m.Match(cmux.HTTP2())
-		// All other connections will get the default treatment.
-		anyL := m.Match(cmux.Any())
-
-		// Construct our h2c handler function.
-		var h2 http2.Server
-		serveConnOpts := &http2.ServeConnOpts{
-			Handler: handler,
-		}
-		serveH2 := func(conn net.Conn) {
-			h2.ServeConn(conn, serveConnOpts)
-		}
-
-		// Start serving HTTP1 on all non-HTTP2 connections.
-		serveConn := ServeHandler(stopper, handler, anyL, tlsConfig)
-
-		// Start serving h2c on all HTTP2 connections.
-		stopper.RunWorker(func() {
-			FatalIfUnexpected(serveConn(h2L, serveH2))
-		})
-
-		// Finally start the multiplexing listener.
-		stopper.RunWorker(func() {
-			FatalIfUnexpected(m.Serve())
-		})
-	}
+	stopper.RunWorker(func() {
+		FatalIfUnexpected(server.Serve(ln))
+	})
 	return ln, nil
 }
 
@@ -162,10 +127,12 @@ func ServeHandler(stopper *stop.Stopper, handler http.Handler, ln net.Listener, 
 	}
 }
 
-// IsClosedConnection returns true if err is cmux.ErrListenerClosed, io.EOF, or
-// the net package's errClosed.
+// IsClosedConnection returns true if err is cmux.ErrListenerClosed,
+// grpc.ErrServerStopped, io.EOF, or the net package's errClosed.
 func IsClosedConnection(err error) bool {
-	return err == cmux.ErrListenerClosed || err == io.EOF ||
+	return err == cmux.ErrListenerClosed ||
+		err == grpc.ErrServerStopped ||
+		err == io.EOF ||
 		strings.Contains(err.Error(), "use of closed network connection")
 }
 

